app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/casiogb6900/CasioGB6900DeviceSupport.java

```diff
@@ -0,0 +1,528 @@
+                    writeCasioVirtualServerFeature(builder);
+                    performImmediately(builder);
+                    handled = true;
+                } catch (IOException e) {
+                    LOG.warn(e.getMessage());
+                }
+                break;
+        }
+        return handled;
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt,
+                                           BluetoothGattCharacteristic characteristic) {
+        boolean handled = false;
+
+        if (super.onCharacteristicChanged(gatt, characteristic)) {
+            return true;
+        }
+
+        UUID characteristicUUID = characteristic.getUuid();
+        byte[] data = characteristic.getValue();
+        if (data.length == 0)
+            return true;
+
+        if(characteristicUUID.equals(CasioGB6900Constants.CASIO_A_NOT_W_REQ_NOT))
+        {
+            handled = handleCasioCom(data);
+        }
+
+        if(characteristicUUID.equals(CasioGB6900Constants.CASIO_A_NOT_COM_SET_NOT))
+        {
+            handled = handleCasioCom(data);
+        }
+
+        if(characteristicUUID.equals(CasioGB6900Constants.ALERT_LEVEL_CHARACTERISTIC_UUID))
+        {
+            GBDeviceEventFindPhone findPhoneEvent = new GBDeviceEventFindPhone();
+            if(data[0] == 0x02) {
+                findPhoneEvent.event = GBDeviceEventFindPhone.Event.START;
+            }
+            else
+            {
+                findPhoneEvent.event = GBDeviceEventFindPhone.Event.STOP;
+            }
+                evaluateGBDeviceEvent(findPhoneEvent);
+            handled = true;
+        }
+
+        if(characteristicUUID.equals(CasioGB6900Constants.RINGER_CONTROL_POINT))
+        {
+            if(data[0] == 0x02)
+            {
+                LOG.info("Mute/ignore call event not yet supported by GB");
+            }
+            handled = true;
+        }
+
+        if(!handled)
+        {
+            LOG.info("Unhandled characteristic change: " + characteristicUUID + " code: " + String.format("0x%1x ...", data[0]));
+        }
+        return true;
+    }
+
+    private void showNotification(byte icon, String title, String message) {
+        try {
+            TransactionBuilder builder = performInitialized("showNotification");
+            int len;
+
+            byte[] titleBytes = title.getBytes(StandardCharsets.US_ASCII);
+            len = titleBytes.length > 18 ? 18 : titleBytes.length;
+            byte[] msg = new byte[2 + len];
+            msg[0] = icon;
+            msg[1] = 1;
+            for(int i=0; i<len; i++)
+            {
+                msg[i + 2] = titleBytes[i];
+            }
+
+            builder.write(getCharacteristic(CasioGB6900Constants.ALERT_CHARACTERISTIC_UUID), msg);
+            LOG.info("Showing notification, title: " + title + " message (not sent): " + message);
+            performConnected(builder.getTransaction());
```

Comment:
As far as I understood the logic, it works the following:
  - Create a new transaction using performInitialized, this makes sure the device is initialized
  - Add actions to the builder
  - Run the builder using performConnected
```
TransactionBuilder builder = performInitialized("showNotification");
builder.write(....);
performConnected(builder.getTransaction());
```
Am I missing something here?

Reply:
I'm going to add some javadoc to these methods and also change the `HPlusSupport` class, which may be the origin of this (I've seen this in a couple of DeviceSupport classes now).

Until I've done this, here's how it works:
- the `TransactionBuilder` is used to *collect* a number of BLE interaction sequences
- once the interaction sequence is complete, one lets the builder create a `Transaction`with all the individual reads, writes, notifys, etc.
- this `Transaction` will get executed by `BtLEQueue`, which takes care of performing the reads, writes etc. in the correct order with the proper synchronization.

The performXXX() methods are just a way to perform something *before* the transaction is executed. I.e. performConnected() will make sure that the device is connected before running the transaction. 

performInitialized() will take care that the device is not only connected, but that the whole initialization is already done. The reason is "autoconnect". If you have a device that is *not connected*, and you receive a message (e.g. a mail, sms, ...), you need to first connect to the device, perform the entire initialization sequence and only then send the message to the device for display, vibration or whatever.

So all you do is
```
TransactionBuilder builder = performInitialized("showNotification");
builder.write(....);
builder.queue(getQueue());
```

If you want to interact with the device without sending the initialization sequence in advance (e.g. because your are part of the initialization sequence), you may use `performConnected()`.

`performImmediately()` is a *very* special thing and rarely needed. It will bypass the ordering of the currently executed transaction in order to perform some BLE interactions *immediately*.

Reply:
Oh, i see, thanks for clarifying the situation! Indeed, I took parts of the code from other device support classes, but I don't remember which one, since I went through most of them.

In this case, performImmediately() shouldn't be needed. I'll push the updated code soon.